pkg/api/endpoints/util.go

```diff
@@ -34,70 +34,93 @@ import (
-				}
-				portsToAddrs[*epp].Insert(allAddrs[ak])
+				mapAddressByPort(&subsets[i].Addresses[k], port, true, allAddrs, portsToAddrs)
+			}
+			for k := range subsets[i].NotReadyAddresses {
+				mapAddressByPort(&subsets[i].NotReadyAddresses[k], port, false, allAddrs, portsToAddrs)
 			}
 		}
 	}
 
 	// Next, map the sets of hosts to the sets of ports they offer.
 	// Go does not allow maps or slices as keys to maps, so we have
-	// to synthesize and artificial key and do a sort of 2-part
+	// to synthesize an artificial key and do a sort of 2-part
 	// associative entity.
 	type keyString string
 	addrSets := map[keyString]addressSet{}
 	addrSetsToPorts := map[keyString][]api.EndpointPort{}
-	for epp, addrs := range portsToAddrs {
+	for port, addrs := range portsToAddrs {
 		key := keyString(hashAddresses(addrs))
 		addrSets[key] = addrs
-		addrSetsToPorts[key] = append(addrSetsToPorts[key], epp)
+		addrSetsToPorts[key] = append(addrSetsToPorts[key], port)
 	}
 
 	// Next, build the N-to-M association the API wants.
 	final := []api.EndpointSubset{}
 	for key, ports := range addrSetsToPorts {
-		addrs := []api.EndpointAddress{}
-		for k := range addrSets[key] {
-			addrs = append(addrs, *k)
+		var addrs, notReadyAddrs []api.EndpointAddress
+		for addr, ready := range addrSets[key] {
+			if ready {
+				addrs = append(addrs, *addr)
+			} else {
+				notReadyAddrs = append(notReadyAddrs, *addr)
+			}
 		}
-		final = append(final, api.EndpointSubset{Addresses: addrs, Ports: ports})
+		final = append(final, api.EndpointSubset{Addresses: addrs, NotReadyAddresses: notReadyAddrs, Ports: ports})
 	}
 
 	// Finally, sort it.
 	return SortSubsets(final)
 }
 
-type addressSet map[*api.EndpointAddress]struct{}
+// The sets of hosts must be de-duped, using IP+UID as the key.
+type addressKey struct {
+	ip  string
+	uid types.UID
+}
+
+// mapAddressByPort adds an address into a map by its ports, registering the address with a unique pointer, and preserving
+// any existing ready state.
+func mapAddressByPort(addr *api.EndpointAddress, port api.EndpointPort, ready bool, allAddrs map[addressKey]*api.EndpointAddress, portsToAddrs map[api.EndpointPort]addressSet) *api.EndpointAddress {
+	key := addressKey{ip: addr.IP}
+	if addr.TargetRef != nil {
+		key.uid = addr.TargetRef.UID
+	}
+
+	// Accumulate the address.
+	existingAddress := allAddrs[key]
+	if existingAddress == nil {
+		// Make a copy so we don't write to the
+		// input args of this function.
+		existingAddress = &api.EndpointAddress{}
+		*existingAddress = *addr
+		allAddrs[key] = existingAddress
+	}
+
+	// Remember that this port maps to this address.
+	if _, found := portsToAddrs[port]; !found {
+		portsToAddrs[port] = addressSet{}
+	}
+	// if we have not yet recorded this port for this address, or if the previous
+	// state was ready, write the current ready state. not ready always trumps
+	// ready.
+	if wasReady, found := portsToAddrs[port][existingAddress]; !found || wasReady {
```

Comment:
I think we should use addressKey as key, instead of existingAddress
Two existingAddress structs can be the same logically, but not be the same by go's perspective.